### DISCUSSION TITLE
Remove files that are no longer in the source directory?

### DISCUSSION THREAD

**Discussion by poperigby** [twpayne/chezmoi#1446] [Q&A] [answered]
I have some files that I deleted in the source directory, but when I look in the target directory, they're still there. I thought `chezmoi apply` would fix this, but it didn't.

**Comment by twpayne** [comment, marked as answer]
chezmoi only looks at the current state of the source directory, it doesn't know anything about its history, and so can't tell if a file has been removed.

To get chezmoi to remove files, you need to create a [`.chezmoiremove` file](https://github.com/twpayne/chezmoi/blob/master/docs/HOWTO.md#ensure-that-a-target-is-removed).

chezmoi takes particular care to make sure that it only removes files when explicitly requested by the user to reduce the chance of inadvertently deleting files.

**Reply by poperigby** [reply]
Thank you!

**Reply by Crocmagnon** [reply]
The link is obsolete.

Moved to : https://github.com/twpayne/chezmoi/blob/master/assets/chezmoi.io/docs/reference/special-files-and-directories/chezmoiremove.md

**Reply by gschintgen** [reply]
and now it's here ;-)
https://github.com/twpayne/chezmoi/blob/master/assets/chezmoi.io/docs/reference/special-files/chezmoiremove.md

**Reply by barraIhsan** [reply]
I know this is many years ahead but..
> it doesn't know anything about its history

can it *somehow* integrate with git? and see idk, maybe get the patch file and we now know that the file is being deleted?

**Reply by halostatue** [reply]
> I know this is many years ahead but..
> 
> > it doesn't know anything about its history
> 
> can it *somehow* integrate with git? and see idk, maybe get the patch file and we now know that the file is being deleted?

chezmoi has git integration, but does not depend on git in operations, unless you use an external git repo.

It will work with fossil, mercurial, pijul or even no repo at all (the latter isn't recommended).

**Reply by barraIhsan** [reply]
> chezmoi has git integration

yea, what I meant is that, with version control we can tell if a file has been removed. So maybe, we can retrieve that info? But only when the user uses version control

**Reply by twpayne** [reply]
This cannot work because it does not distinguish between the cases "I removed this file from version control because I want to chezmoi to remove it on all my machines" and "I removed this file from version control because I no longer want chezmoi to manage this file".

This is why [.chezmoiremove](https://www.chezmoi.io/reference/special-files/chezmoiremove/) exists.

**Reply by barraIhsan** [reply]
Wait, yeah that make sense.

**Reply by Lockszmith-GH** [reply]
Sorry to re-necromance this thread, @twpayne .

Would it make sense to have the [**forget**](https://www.chezmoi.io/reference/commands/forget/) and [**destroy**](https://www.chezmoi.io/reference/commands/destroy/) commands have a flag (let's say `--sync` or `--managed`) that makes sure the source/working-copy is updated with a `remove_` [source-state-attributed](https://www.chezmoi.io/reference/source-state-attributes/) blank-file?

This would practically mean: 'remove on my other systems on next sync`

**Reply by barraIhsan** [reply]
And sometimes we just want to rename a file here and there. And adding to chezmoidelete would be *somewhat inconvenient*

**Reply by Lockszmith-GH** [reply]
Not sure what you mean @barraIhsan.

Renaming a file 'everywhere chezmoi works' would have to mean: manage the old file name as removed and then add the new files. I don't see anyway around that.

Maybe an `chemzoi mv` command that will automatically create a `remove_` source-state would probably be a 'more automated' way.

Another idea I have would be to have the `remove_` files contain a timestamp, either a creation time, or an expiration time.
This will allow me to write a garbage-collection type script to remove all of the `remove_` files "older than a month" form the source git repo.

**Reply by barraIhsan** [reply]
what i meant is I don't want a massive list of filename sitting on chezmoiremove just to make sure they removed properly.

Perhaps, your idea could solve this, give it some time and then remove the filename entry from the chezmoiremove

**Reply by bradenhilton** [reply]
@barraIhsan I've yet to be corrected, so I will assume that the method I described in https://github.com/twpayne/chezmoi/discussions/4243#discussioncomment-12007027 does not contain any errors.

You can see in my comment that I mention removing entries from `.chezmoiremove` once the change has propagated to all of your machines. You don't need to keep entries in there forever.

**Reply by barraIhsan** [reply]
The issue here is with version control. Suppose we add a file entry to chezmoiremove in one commit and then remove it in the next.

If someone is using our dotfiles, they might have skipped that specific commit.

For example, if you added the file entry in `HEAD~3` and removed it in `HEAD~2`, then continued making regular changes, a user who is currently on `HEAD~5` will pull the latest changes directly from `HEAD`. This means they’ll bypass the intermediate commits, potentially missing the removal of the file entry.

**Reply by bradenhilton** [reply]
Of course, you are correct, that's why I took care to specify that an entry should only be removed when you are sure the change has propagated everywhere it is needed.

**Reply by Lockszmith-GH** [reply]
I've created #4361 FR, hopefully it will pique some interest to develop it.

**Comment by rbhanot4739** [comment]
Since chezmoi target directory is internally a git repo so an alternate solution I tend to follow is to use the `git rm` to delete the files from chezmoi target directory which no longer exists in source state.

```
❯ chezmoi status
DA .config/nvim/lua/plugins/chezmoi.lua
DA .config/nvim/lua/plugins/lsp-tools.lua
DA .config/nvim/lua/plugins/refector.lua


❯ git rm dot_config/nvim/lua/plugins/chezmoi.lua
rm 'dot_config/nvim/lua/plugins/chezmoi.lua'
❯ chezmoi status
DA .config/nvim/lua/plugins/lsp-tools.lua
DA .config/nvim/lua/plugins/refector.lua
```

**Reply by Lockszmith-GH** [reply]
right, but that will not remove them from another machine after you did a git pull there.
That's where the remove_... files are required.
